Clarify behavior of Span.End, interaction with IsRecording (#1011)
open-telemetry/opentelemetry-specification: specification/trace/api.md

`IsRemote` to true on a parent `SpanContext` so `Span` creation knows if the
parent is remote.

> Any span that is created MUST also be ended, usually by calling [`End`](#end)

Reviewer: Usually?

Author: Some SIGs provide convenience mechanisms such as putting the span in some scope/block after which it is ended automatically (e.g. https://github.com/open-telemetry/opentelemetry-python/blob/c1ec4445719e3872366f172443de53fd12640d07/opentelemetry-api/src/opentelemetry/trace/__init__.py#L290)

Reviewer 2: That still calls `End` though, right?

This addition seems more like a documentation point to make clear to users if they use a certain way of starting a span they must call `end`. There is no way that the API or SDK can ensure that all spans are ended so a `MUST` here is problematic.

Maybe it should be that an API/SDK `SHOULD` provide a way to use spans without needing explicit ends so that they can't accidentally leak spans -- like scopes/blocks.

Author: > That still calls End though, right?

That is hidden away in the API in that case. Some implementation might not call end.

> There is no way that the API or SDK can ensure that all spans are ended so a MUST here is problematic.

Rather than ensure that it happens, the intent here is that the SDK can rely on the user ensuring that it happens. Which is clarified by the following MAY.

Reviewer: I suggest removing this part, as as @tsloughter said, we still call `End()` (even if indirectly). It should be clear to users that `End()` is called in all cases IMHO (or at least have an equivalent semantic behavior, such as an implementation calling `EndInternal()`).

Otherwise, I'm afraid that this means we are opening the door for adding other methods to `End()` the `Span`.

Author: > Otherwise, I'm afraid that this means we are opening the door for adding other methods to End() the Span.

This ship has already sailed in Python. We could have them do a breaking change here though to move that convenience method out of the API, so that the implementation will have to call End. Currently, API implementations may use other means to end the span, if `start_as_current_span` is used in Python.

CC @open-telemetry/python-approvers @toumorokoshi

Reviewer 3: The convenience methods [still call end](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L817), and are theoretically in the control of the user. 

So I would argue that Python still accomplishes the *intent* of the user calling end(). It's also documented as calling end(), so I would +1 [Yuri's suggestion](https://github.com/open-telemetry/opentelemetry-specification/pull/1011/files#r495042795).

We should allow APIs that are natural in the language. Not every language exposes contextmanagers, but those where it's idiomatic should be able to use them.

Reviewer 2: I just read the Python comment linked to and I think it is wrong? Though the implementation may be correct:

> On exiting the context manager stops the span and set its parent as the current span.

It should be setting the previously active span as active, not the parent which could be set to anything with starting a span. Based on the comment `# returns previously active span` on the example I take it the actual implementation does not simply set the parent to the current span but it instead returns to the previous context.

And yea, this is the same as we are doing in Erlang but we don't have `with`.

Author: > The convenience methods still call end,

Only in the SDK implementation, but a different implementation might not call end. I think the important think is that the span is ended, not that end is called. If the API has another method that is documented to end the span, is it really important for all implementations to call exactly end()?

Author: > I just read the Python comment linked to and I think it is wrong? Though the implementation may be correct:
>
>> On exiting the context manager stops the span and set its parent as the current span.

Yeah, if it really does that, this seems like a pitfall.

Reviewer 3: The documentation needs some updates, to your point @tsloughter it just pops the context.

[Here's a pr](https://github.com/open-telemetry/opentelemetry-python/pull/1170) to addresses those. I'm confident my comments are still correct in terms of the *behavior*.

Author: In terms of behavior, the span is ended. In terms of implementation, end is actually called. Do we need to specify the implementation of convenience methods?

Reviewer 3: Theoretically the spec could call out that span.End() is the *only* way to end a span, and that would imply that convenience methods must call span.End(). 

I'm not strongly opinionated but I agree with @carlosalberto that ensuring span.End() is the only way to end a span is very helpful from a standardization / debugging standpoint (no need to be aware of some exception in a specific language's implementation).

Author: > I agree with @carlosalberto that ensuring span.End() is the only way to end a span

Then start_as_current_span violates this, because it is a different API that also ends the span. I will try to come up with a better wording.

Author: I added 4389e48. What do you think?

Reviewer 2: How is a span ended when `start_as_current_span` is used?

Reviewer 2: I ask in part because I've been intending to base some of the Erlang API updated on the Python API since they appear fairly similar.

Author: A span is ended by calling it's end even with start_as_current_span. See here https://github.com/open-telemetry/opentelemetry-python/blob/c1ec4445719e3872366f172443de53fd12640d07/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L701-L710
and here https://github.com/open-telemetry/opentelemetry-python/blob/c1ec4445719e3872366f172443de53fd12640d07/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L787-L817 Note that the part after the `yield` is executed after the `with` block in which the returned thing is used is exited.

Reviewer 2: Ah, thanks, I misread and thought you said it did not call `span.end()`.

And good to know the `start_as_current_span` relies on `yield` and `with`.

> Any span that is created MUST also be ended, usually by calling [`End`](#end)

Reviewer 4: ```suggestion
Any span that is created MUST also be ended by the instrumentation, by calling [`End`](#end) either directly or indirectly through special with-resource language features (e.g. `with span:` in Python).
```

Author: Please see thread above: https://github.com/open-telemetry/opentelemetry-specification/pull/1011#discussion_r494904498

`AddEvent` operation, attributes using `SetAttributes`, status with `SetStatus`,
etc.

> Note that after a span is ended, the Span usually becomes non-recording and thus

Reviewer 2: Is this really what `is_recording` should be doing? I'm not against it, just surprised and didn't consider it the role of `is_recording`.

Author: I thought if the span is ended, it does not make sense to compute attributes for it any longer, so for the use-cases of IsRecording that I can think of, it does make sense. It might be surprising if an exporter/OnEnd SpanProcessor checks IsRecording, as they are invoked *after* ending the span.

Reviewer: I'm fine with this changing, but wonder if @bogdandrutu has anything to say about this part (disclaimer: we didn't have this in OT, so it's mostly an OC notion).

Author: IMHO, IsRecording should answer the question "Does it make sense to compute anything and set it as attribute now?". And for an ended span, the answer should be "no".

Reviewer: Ping @bogdandrutu ;)

Reviewer 2: Still no comment from @bogdandrutu. We should discuss this in the spec meeting today.

Author: Discussed in the SIG mtg: It should be fine to have this as a "SHOULD" requirement. There may be implementations were synchronization is problematic (although synchronization of other stuff is already required) or expensive -- as IsRecording is meant for improving performance, implementations should be allowed to continue returning true after end if changing the return value would be too expensive.

Author: Updated in 0f98846. I did not explicitly mention synchronization though, as I don't think this will be a common concern and also the next paragraph after the next already states "This flag SHOULD be used to avoid expensive computations" so hopefully this is clear enough to implementers.

IsRecording should consequently return false for ended Spans.

> There SHOULD be no parameter.

Reviewer 4: ```suggestion
`IsRecording` SHOULD not take any parameters.
```

#### End

> Finish the `Span`. This call will take the current timestamp to set as `Span`'s

Reviewer 4: ```suggestion
End the `Span`. This call will take the current timestamp to set as `Span`'s
```

Author: Still wrong. Not really related to this PR, but I will push a fix...

Author: Actually, it *is* related, considering the title 😄

Author: What do you think of 453bc78? I hope I did not go overboard?
